{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perry Wasdin\n",
    "05/31/24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concatenating and filtering LIBRA-seq outputs from published datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['200301_1655331_hiv_flu_N26_epmappingpaper_AS.xlsx',\n",
       " '200309_150517_hiv_flu_N27_epmappingpaper_AS.xlsx',\n",
       " '200324_4647-1_cov_naivedonor_AS.xlsx',\n",
       " '200324_4647-2_cov_sarsdonor_crmpaper_AS.xlsx',\n",
       " '201022_5317_sars2_potencypaper_AS.xlsx',\n",
       " '201022_5318-1_sars2_potencypaper_AS.xlsx',\n",
       " '201022_5318-2_sars2_potencypaper_AS.xlsx',\n",
       " '201113_4591-1_hiv_epmappingmaper_RV.xlsx',\n",
       " '210316_5885_sars1sars2_doubleblocker_potencypaper_AS.xlsx',\n",
       " '210616_6420_hiv_epmappingpaper_LW.xlsx',\n",
       " '210619_6421-1_hiv_epmapping_LW.xlsx',\n",
       " '210629_6421-2_hiv_epmapping_LW.xlsx',\n",
       " '211114_7128-1_cov_pooledpeds_SW.csv',\n",
       " '211119_7128-2_cov_pooledpeds_SW.csv',\n",
       " 'concat_5404_paired_chains_24-03-06.csv',\n",
       " 'KK_5812_concat_LIBRA-seq_outputs_PW_24-02-13.csv']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_data_dir = './raw_data/'\n",
    "raw_files = [i for i in os.listdir(raw_data_dir) if ('.csv' in i) or ('.xlsx' in i)]\n",
    "raw_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "hiv_datasets = ['1655331', '150517', '4591-1', '6420', '6421-1', '6421-2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get binders, returning DF with barcode, antigen name, LSS score, and UMI counts\n",
    "def get_binders(lss_col, lss_thresh, umi_thresh):\n",
    "    antigen = lss_col.split('.LSS')[0]\n",
    "    umi_col = antigen\n",
    "    if umi_col not in df.columns:\n",
    "        umi_col = umi_col + '.UMI'\n",
    "\n",
    "    binder_df = df[[lss_col, umi_col]]\n",
    "\n",
    "    binder_df = binder_df[binder_df[lss_col] >= lss_thresh] \n",
    "    binder_df = binder_df[binder_df[umi_col] >= umi_thresh] \n",
    "\n",
    "    # binder_df.index = antigen + ':' + binder_df.index\n",
    "    binder_df.columns = ['LSS', 'UMI']\n",
    "    binder_df['Antigen'] = antigen\n",
    "\n",
    "    return binder_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_cols = ['SEQUENCE_INPUT.H', 'V_CALL.H', 'D_CALL.H', 'J_CALL.H',\n",
    "       'JUNCTION.H', 'V_IDENTITY.H', 'J_IDENTITY.H', 'FWR1_IMGT.H', 'FWR2_IMGT.H', 'FWR3_IMGT.H',\n",
    "       'FWR4_IMGT.H', 'CDR1_IMGT.H', 'CDR2_IMGT.H', 'CDR3_IMGT.H', 'C_CALL.H',\n",
    "       'SEQUENCE_INPUT.L', 'V_CALL.L', 'J_CALL.L', 'SEQUENCE_VDJ.L',\n",
    "       'JUNCTION.L', 'V_IDENTITY.L', 'J_IDENTITY.L', 'FWR1_IMGT.L',\n",
    "       'FWR2_IMGT.L', 'FWR3_IMGT.L', 'FWR4_IMGT.L', 'CDR1_IMGT.L',\n",
    "       'CDR2_IMGT.L', 'CDR3_IMGT.L', 'C_CALL.L', 'D_CALL.L', 'N']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "215 binders for Sample 1655331\n",
      "2680 binders for Sample 150517\n",
      "1773 binders for Sample 4647-1\n",
      "2313 binders for Sample 4647-2\n",
      "449 binders for Sample 5317\n",
      "357 binders for Sample 5318-1\n",
      "579 binders for Sample 5318-2\n",
      "2937 binders for Sample 4591-1\n",
      "1814 binders for Sample 5885\n",
      "359 binders for Sample 6420\n",
      "56 binders for Sample 6421-1\n",
      "no IgG for 6421-2\n",
      "0 binders for Sample 7128-1\n",
      "3015 binders for Sample 7128-2\n",
      "5144 binders for Sample 5404\n",
      "4305 binders for Sample 5812\n"
     ]
    }
   ],
   "source": [
    "binding_dfs = {}\n",
    "all_data = {}\n",
    "\n",
    "for i in raw_files:\n",
    "\n",
    "    # Read in file from either CSV or Excel spreadsheet\n",
    "    if ('.csv' in i):\n",
    "        df = pd.read_csv(raw_data_dir + i, index_col='BARCODE')\n",
    "\n",
    "    elif ('.xlsx' in i):\n",
    "        df = pd.read_excel(raw_data_dir + i, index_col='BARCODE')\n",
    "\n",
    "    id = i.split('_')[1].split('_')[0] # Get VANTAGE ID\n",
    "\n",
    "    # Check if IgG in sample\n",
    "    if not ('IGHG1' in df['C_CALL.H'].unique()):\n",
    "        print('no IgG for ' + id)\n",
    "        continue\n",
    "\n",
    "    # Select isotypes to remove\n",
    "    # df = df[~df['C_CALL.H'].isin(['IGHM', 'IGHD'])]\n",
    "    \n",
    "    df[df['N'] == 1] # Only keep cells with 1 heavy chain\n",
    "\n",
    "    if not (len(df) > 5): # Skip samples with very low cell counts\n",
    "        print('low cell count for ' + id)\n",
    "        continue\n",
    "\n",
    "    df.index = df.index + '_' + np.array(range(len(df))).astype(str)\n",
    "\n",
    "    lss_names = [i for i in df.columns if '.LSS' in i]\n",
    "\n",
    "    # Get binding cells with function defined above, based on LSS and UMI thresholds\n",
    "    binders = pd.concat([get_binders(col, 1, 20) for col in lss_names])\n",
    "\n",
    "    print (str(len(binders)) + ' binders for Sample ' + str(id))\n",
    "\n",
    "    binding_dfs[id] = binders\n",
    "    all_data[id] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_df = pd.concat(binding_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "binding_groups = {'HIV':['HIV_ZM197', 'HIV_CZA97', 'HIV_A244', 'HIV_ConC', 'HIV_CNE55', 'BG505_6R_N332', 'CZA97_6R_N332', 'BG505sc_N332T', 'CZA97sc_N332T', 'BG505sc_N160K', 'CZA97sc_N160K', \n",
    "                         'BG505sc_K169E', 'CZA97sc_K169E', 'BG505sc_D279K', 'CZA97sc_D279K', 'BG505sc_D368R', 'CZA97sc_D368R', 'BG505sc_DKO', 'CZA97sc_DKO', 'HIV_BG505_D279K', 'HIV_BG505_D368R', \n",
    "                         'HIV_BG505_K169E', 'HIV_BG505_N160K', 'HIV_BG505_N332T', 'HIV_BG505_DKO', 'HIV_BG505_T332N'],\n",
    "\n",
    "                'CoV':['COV_HKU1', 'COV_MERS', 'COV_SARS', 'COV_S1Fd', 'COV_OC43', 'COV_SARS2', 'CoV_HP6', 'CoV_HP1', 'CoV_SARS2', 'CoV_SARS1', 'SARS2', 'SARS1', 'MERS', 'OC43', '229E', 'SARS2_SA', 'SARS2_UK', \n",
    "                'RaTG13', 'Civet_007_2004', 'CoV_MERS', 'CoV_OC43', 'CoV_NL63', 'CoV_229E', 'CoV_SARS2D614G', 'CoV_SARS2NTD', 'CoV_SARS1RBD', 'CoV_MERSRBD', 'CoV_HKU1'],\n",
    "\n",
    "                'Flu':['FLU_HA_H9', 'FLU_HA_H10', 'FLU_HA_Indo', 'FLU_HA_Anhui', 'FLU_HA_Michigan', 'FLU_HANC99', 'HA_NC99', 'FLU_HA_NC99'],\n",
    "\n",
    "                'HCV': ['JFH1_E2']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_antigens = np.concatenate(list(binding_groups.values()))\n",
    "concat_df = concat_df[concat_df['Antigen'].isin(all_antigens)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_keys_by_value_in_list(my_dict, search_value):\n",
    "    return [key for key, value_list in my_dict.items() if search_value in value_list][0]\n",
    "\n",
    "def flag_polyreactive(antigen_list):\n",
    "    bound_groups = [find_keys_by_value_in_list(binding_groups, ant) for ant in antigen_list]\n",
    " \n",
    "    if len(np.unique(bound_groups)) > 1:\n",
    "        return 'Polyreactive'\n",
    "    else:\n",
    "        return 'Specific'\n",
    "    \n",
    "def get_group(antigen):\n",
    "    bound_groups = [find_keys_by_value_in_list(binding_groups, ant) for ant in [antigen]]\n",
    "    return bound_groups[0]\n",
    "\n",
    "antigens_bound = concat_df.groupby(concat_df.index)['Antigen'].unique()\n",
    "antigens_bound['Reactivity'] = antigens_bound.apply(flag_polyreactive)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CoV    16856\n",
       "HIV     6408\n",
       "Flu     1233\n",
       "HCV       14\n",
       "Name: Group, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_df['Group'] = concat_df['Antigen'].apply(get_group)\n",
    "concat_df['Group'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most of the time we remove polyreactive cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Specific        22418\n",
       "Polyreactive     2093\n",
       "Name: Reactivity, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_df['Reactivity'] = antigens_bound['Reactivity']\n",
    "concat_df['Reactivity'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_df = concat_df[concat_df['Reactivity'] != 'Polyreactive']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_df['Vantage_ID'] = concat_df.index.get_level_values(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For non-HIV datasets, we assume HIV binding antibodies are false positives (or non-specific binding)\n",
    "def label_neg_control(row):\n",
    "    if (row['Antigen'] in (binding_groups['HIV'])) and (row['Vantage_ID'] not in (hiv_datasets)):\n",
    "        return 'neg control'\n",
    "    else:\n",
    "        return 'binding'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "binding        21674\n",
       "neg control      744\n",
       "Name: neg_control, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_df['neg_control'] = concat_df.apply(label_neg_control, axis=1)\n",
    "concat_df['neg_control'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_df = concat_df[concat_df['neg_control'] != 'neg control']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_data_concat = pd.concat(all_data).loc[concat_df.index]\n",
    "\n",
    "# Some of the sequences are missing the raw input sequences. For purposes of getting the variable chains, we can just infill\n",
    "# these with teh variable sequences assigned by IMGT\n",
    "\n",
    "meta_data_concat['SEQUENCE_INPUT.H'] = meta_data_concat['SEQUENCE_INPUT.H'].fillna(meta_data_concat['SEQUENCE_VDJ.H'])\n",
    "meta_data_concat['SEQUENCE_INPUT.L'] = meta_data_concat['SEQUENCE_INPUT.L'].fillna(meta_data_concat['SEQUENCE_VDJ.L'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concat_df['VH_nuc'] = meta_data_concat['SEQUENCE_INPUT.H']\n",
    "# concat_df['VL_nuc'] = meta_data_concat['SEQUENCE_INPUT.L']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_df.index = concat_df.index.get_level_values(0) +':' + concat_df.index.get_level_values(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By default, the LIBRA-seq pipeline outputs nucleotide sequences. We align these to a reference genome and annotate using IMGT HighV-Quest (https://www.imgt.org/HighV-QUEST/home.action) for which you need fasta inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_to_fasta(input_csv, output_fasta_name): # \n",
    "    output = input_csv.copy()\n",
    "\n",
    "    output.index = '>' + output.index\n",
    "\n",
    "    hchains = pd.DataFrame(index=output.index)\n",
    "    hchains['Seq'] = output['VH_nuc']\n",
    "    hchains.index = hchains.index + ' HC'\n",
    "\n",
    "    lchains = pd.DataFrame(index=output.index)\n",
    "    lchains['Seq'] = output['VL_nuc']\n",
    "    lchains.index = lchains.index + ' LC'\n",
    "\n",
    "    output = pd.concat([hchains, lchains])\n",
    "\n",
    "\n",
    "    path = output_fasta_name + '.fasta'\n",
    "\n",
    "    with open(path, \"w\") as f:\n",
    "\n",
    "        for i in range(len(output)):\n",
    "            feature = i + 1\n",
    "            header = output.iloc[i].name\n",
    "            line1 = output.iloc[i][0]\n",
    "\n",
    "            # break\n",
    "            f.write(header + '\\n')\n",
    "\n",
    "            # print('line1 + '\\n')\n",
    "            f.write(line1 + '\\n')\n",
    "\n",
    "write_to_fasta(concat_df, 'published_LIBRA-seq_binders_genV2_24-03-06')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concat_df.to_csv('published_LIBRA-seq_binders_genV2_24-03-06.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "imgt_aa_seqs = pd.read_csv('./published_LIBRA_seq_binders_genV2_24_03_06_IMGT/5_AA-sequences.txt', sep='\\t') # Read output from IMGT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence number</th>\n",
       "      <th>Sequence ID</th>\n",
       "      <th>V-DOMAIN Functionality</th>\n",
       "      <th>V-GENE and allele</th>\n",
       "      <th>J-GENE and allele</th>\n",
       "      <th>D-GENE and allele</th>\n",
       "      <th>V-D-J-REGION</th>\n",
       "      <th>V-J-REGION</th>\n",
       "      <th>V-REGION</th>\n",
       "      <th>FR1-IMGT</th>\n",
       "      <th>CDR1-IMGT</th>\n",
       "      <th>FR2-IMGT</th>\n",
       "      <th>CDR2-IMGT</th>\n",
       "      <th>FR3-IMGT</th>\n",
       "      <th>CDR3-IMGT</th>\n",
       "      <th>JUNCTION</th>\n",
       "      <th>J-REGION</th>\n",
       "      <th>FR4-IMGT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1655331:AGTTGGTAGTCTTGCA_11_HC</td>\n",
       "      <td>productive</td>\n",
       "      <td>Homsap IGHV4-4*07 F, or Homsap IGHV4-4*12 F</td>\n",
       "      <td>Homsap IGHJ3*02 F</td>\n",
       "      <td>Homsap IGHD2-15*01 F</td>\n",
       "      <td>QMQLQESGPGLVKSSETLSLTCSVSGDSIDGYYWSWIRQAADKRLE...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QMQLQESGPGLVKSSETLSLTCSVSGDSIDGYYWSWIRQAADKRLE...</td>\n",
       "      <td>QMQLQESGPGLVKSSETLSLTCSVS</td>\n",
       "      <td>GDSIDGYY</td>\n",
       "      <td>WSWIRQAADKRLEWIGR</td>\n",
       "      <td>IHTSGTT</td>\n",
       "      <td>NFNPSLRSRVTMSVDTSENQVSLRLTSVTAADTAVYYC</td>\n",
       "      <td>ARDLVVVLSAREVDDAFDI</td>\n",
       "      <td>CARDLVVVLSAREVDDAFDIW</td>\n",
       "      <td>DAFDIWGQGTTVTVSS</td>\n",
       "      <td>WGQGTTVTVSS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Sequence number                     Sequence ID V-DOMAIN Functionality  \\\n",
       "0                1  1655331:AGTTGGTAGTCTTGCA_11_HC             productive   \n",
       "\n",
       "                             V-GENE and allele  J-GENE and allele  \\\n",
       "0  Homsap IGHV4-4*07 F, or Homsap IGHV4-4*12 F  Homsap IGHJ3*02 F   \n",
       "\n",
       "      D-GENE and allele                                       V-D-J-REGION  \\\n",
       "0  Homsap IGHD2-15*01 F  QMQLQESGPGLVKSSETLSLTCSVSGDSIDGYYWSWIRQAADKRLE...   \n",
       "\n",
       "  V-J-REGION                                           V-REGION  \\\n",
       "0        NaN  QMQLQESGPGLVKSSETLSLTCSVSGDSIDGYYWSWIRQAADKRLE...   \n",
       "\n",
       "                    FR1-IMGT CDR1-IMGT           FR2-IMGT CDR2-IMGT  \\\n",
       "0  QMQLQESGPGLVKSSETLSLTCSVS  GDSIDGYY  WSWIRQAADKRLEWIGR   IHTSGTT   \n",
       "\n",
       "                                 FR3-IMGT            CDR3-IMGT  \\\n",
       "0  NFNPSLRSRVTMSVDTSENQVSLRLTSVTAADTAVYYC  ARDLVVVLSAREVDDAFDI   \n",
       "\n",
       "                JUNCTION          J-REGION     FR4-IMGT  \n",
       "0  CARDLVVVLSAREVDDAFDIW  DAFDIWGQGTTVTVSS  WGQGTTVTVSS  "
      ]
     },
     "execution_count": 188,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imgt_aa_seqs.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_df['VH_AA'] = concat_df.index + '_HC'\n",
    "concat_df['VH_AA'] = concat_df['VH_AA'].map(dict(zip(imgt_aa_seqs['Sequence ID'], imgt_aa_seqs['V-D-J-REGION'])))\n",
    "\n",
    "concat_df['VL_AA'] = concat_df.index + '_LC'\n",
    "concat_df['VL_AA'] = concat_df['VL_AA'].map(dict(zip(imgt_aa_seqs['Sequence ID'], imgt_aa_seqs['V-J-REGION'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "antigen_seqs = pd.read_csv('../antigen_sequences_LIBRA-seq_24-01-08.csv')\n",
    "antigen_seqs = antigen_seqs[~antigen_seqs['Seq'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removed antigens we have low confidence in (from followup analysis)\n",
    "mutated_bg505_ants = ['HIV_BG505_D279K', 'HIV_BG505_D368R',\n",
    "       'HIV_BG505_K169E', 'HIV_BG505_N160K', 'HIV_BG505_N332T',\n",
    "       'HIV_BG505_DKO', 'HIV_BG505_T332N','BG505_6R_N332', 'CZA97_6R_N332', 'BG505sc_N332T', 'CZA97sc_N332T',\n",
    "       'BG505sc_N160K', 'CZA97sc_N160K', 'BG505sc_K169E', 'CZA97sc_K169E',\n",
    "       'BG505sc_D279K', 'CZA97sc_D279K', 'BG505sc_D368R', 'CZA97sc_D368R',\n",
    "       'BG505sc_DKO', 'CZA97sc_DKO', 'CoV_MERS', 'MERS', 'CoV_MERSRBD']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_df = concat_df[~concat_df['Antigen'].isin(mutated_bg505_ants)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1616, 9)"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['CoV_SARS1', 'CoV_SARS2', 'HIV_ConC', 'HIV_A244', 'CoV_SARS2NTD',\n",
       "       'CoV_SARS2D614G', 'CoV_OC43', 'HIV_ZM197', 'HIV_CZA97', 'HIV_CNE55',\n",
       "       'FLU_HANC99', 'FLU_HA_NC99', 'HA_NC99', 'COV_MERS', 'COV_OC43',\n",
       "       'CoV_HKU1', 'COV_SARS', 'COV_SARS2', 'SARS2_UK', 'COV_HKU1', 'CoV_HP6',\n",
       "       'CoV_229E', 'CoV_NL63', 'SARS2_SA', 'FLU_HA_Michigan', 'SARS2',\n",
       "       'RaTG13', 'COV_S1Fd', 'CoV_HP1'],\n",
       "      dtype='object', name='Antigen')"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_df['Antigen'].value_counts()[concat_df['Antigen'].value_counts() >= 5].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Vantage_ID\n",
       "5885    174\n",
       "5404    127\n",
       "5812     21\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 196,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "concat_df[concat_df['Antigen'] == 'CoV_SARS2']['Vantage_ID'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "antigen_conversion = {'HIV_ZM197':'HIV_ZM197_Env', 'HIV_CZA97':'HIV_CZA97_Env', 'HIV_ConC':'ConC', 'HIV_CNE55': 'HIV_CNE55_Env', 'HIV_ZM197':'HIV_ZM197_Env', \n",
    "                      'FLU_HANC99':'HA_H1-NC99', 'FLU_HA_NC99':'HA_H1-NC99', 'HA_NC99':'HA_H1-NC99', 'FLU_HA_Michigan':'H1N1_A_Michigan_45-2015_HA',\n",
    "                     'COV_OC43':'CoV_OC43', 'CoV_OC43':'CoV_OC43', 'CoV_SARS2D614G':'CoV_SARS2D614G', 'CoV_HP6':'SARS2_Hexa',\n",
    "                     'CoV_SARS1': 'CoV_SARS1', 'COV_SARS':'CoV_SARS1', 'COV_SARS2':'SARS2_Hexa', 'CoV_SARS2':'SARS2_Hexa',\n",
    "                     'CoV_SARS2NTD': 'SARS2_NTD', 'CoV_HKU1':'CoV_HKU1', 'COV_HKU1':'CoV_HKU1'\n",
    "                     }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_df['antigen_name'] = concat_df['Antigen'].map(antigen_conversion)\n",
    "concat_df['antigen_seq'] = concat_df['antigen_name'].map(dict(zip(antigen_seqs['Antigen'], antigen_seqs['Seq'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "# concat_df.to_csv('published_LIBRA-seq_binders_genV2_24-03-09.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "GEX2023",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
